src/Avalonia.Controls/Utils/VirtualizingSnapPointsList.cs

```diff
@@ -0,0 +1,191 @@
+                                {
+                                    snapPoint += lastElement.Bounds.Right;
+                                }
+                            }
+                        }
+                        break;
+                    case Orientation.Vertical:
+                        container = _realizedElements.GetElement(index);
+                        if (container != null)
+                        {
+                            switch (_snapPointsAlignment)
+                            {
+                                case SnapPointsAlignment.Near:
+                                    snapPoint = container.Bounds.Top;
+                                    break;
+                                case SnapPointsAlignment.Center:
+                                    snapPoint = container.Bounds.Center.Y;
+                                    break;
+                                case SnapPointsAlignment.Far:
+                                    snapPoint = container.Bounds.Bottom;
+                                    break;
+                            }
+                        }
+                        else
+                        {
+                            var ind = index;
+                            if(index > _realizedElements.LastIndex)
+                            {
+                                ind -= _realizedElements.LastIndex + 1;
+                            }
+                            snapPoint = ind * averageElementSize;
+                            switch (_snapPointsAlignment)
+                            {
+                                case SnapPointsAlignment.Center:
+                                    snapPoint += averageElementSize / 2;
+                                    break;
+                                case SnapPointsAlignment.Far:
+                                    snapPoint += averageElementSize;
+                                    break;
+                            }
+                            if (index > _realizedElements.LastIndex)
+                            {
+                                var lastElement = _realizedElements.GetElement(_realizedElements.LastIndex);
+                                if (lastElement != null)
+                                {
+                                    snapPoint += lastElement.Bounds.Bottom;
+                                }
+                            }
+                        }
+                        break;
+                }
+
+                return snapPoint;
+            }
+        }
+
+        public int Count => _parentOrientation != _orientation ? 0 : _end - _start + 1;
+
+        public IEnumerator<double> GetEnumerator()
+        {
+            return new VirtualizingEnumerator(this);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return new VirtualizingEnumerator(this);
+        }
+
+        private class VirtualizingEnumerator : IEnumerator<double>
```

Comment:
This isn't a criticism, but I'm wondering why a new enumerator class is defined here instead of just using `yield return` in `GetEnumerator()`? Performance?